crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4589,9 +4791,8 @@ mod tests {
         assert_file_diagnostics(&db, "src/a.py", &[]);
 
         // TODO: once we support `sys.version_info` branches,
-        // we can set `--target-version=py311` in this test
-        // and the inferred type will just be `BaseExceptionGroup` --Alex
-        assert_public_ty(&db, "src/a.py", "e", "Unknown | BaseExceptionGroup");
+        // we should set `--target-version=py311` in this test
+        assert_public_ty(&db, "src/a.py", "e", "BaseExceptionGroup");
```

Comment:
I'm not sure this should be changing as part of this PR. It's a class definition, and it's still only defined in the `sys.version_info >= (3, 11)` branch in typeshed: https://github.com/python/typeshed/blob/97ce20034352caff50d818e4b32cc28338a08d19/stdlib/builtins.pyi#L2025-L2032. Since we don't support `sys.version_info` branches yet, we should still be emitting an error complaining about how it's defined in the `if` branch but not the `else` branch.

Reply:
When we import something which has type declarations (and remember, function and class definitions are also declarations), we don't do type inference or look at bindings at all, we just look at the declared type(s).

If a declaration is present in one path and not in another, the current behavior is that we don't care about that; we just union the live declarations, and we never include Unbound. Unbound is a statement about runtime bindings, not a statement about declarations. I guess we could have an equivalent possibly-undeclared type (or piece of metadata in the use-def map), but I'm not sure what we would do with this, other than maybe emit some kind of diagnostic "you are importing something that is not declared in all paths," or similar. I still think we'd want to infer the imported type the way we do now -- I'm not sure that it's an improvement to union the imported type with `Unknown` just because the name wasn't declared in all paths.

Reply:
Open to considering proposed alternative behaviors; just making sure we're clear on what is currently implemented and why.

Reply:
We discussed, and settled on a few modifications to the semantics, which I've now implemented in this PR, and I describe in another comment below. Thanks for prompting a re-think here!